text/3503-frontmatter.md

```diff
@@ -0,0 +1,658 @@
+fn main() {
+    println!("{}", time::now().rfc822z());
+}
+```
+
+## RustExplorer
+
+[Rust Explorer](https://users.rust-lang.org/t/rust-playground-with-the-top-10k-crates/75746)
+uses a comment syntax for specifying dependencies
+
+Example:
+```rust
+/*
+[dependencies]
+actix-web = "*"
+ureq = "*"
+tokio = { version = "*", features = ["full"] }
+*/
+
+use actix_web::App;
+use actix_web::get;
+use actix_web::HttpResponse;
+use actix_web::HttpServer;
+use actix_web::post;
+use actix_web::Responder;
+use actix_web::web;
+use tokio::spawn;
+use tokio::sync::oneshot;
+use tokio::task::spawn_blocking;
+```
+
+## PL/Rust
+
+Example:
+```sql
+CREATE OR REPLACE FUNCTION randint() RETURNS bigint LANGUAGE plrust AS $$
+[dependencies]
+rand = "0.8"
+
+[code]
+use rand::Rng; 
+Ok(Some(rand::thread_rng().gen())) 
+$$;
+```
+
+See [External Dependencies](https://github.com/tcdi/plrust/blob/main/doc/src/dependencies.md)
+
+## YAML frontmatter
+
+As a specialization of [YAML presentation streams](https://yaml.org/spec/1.2.2/#323-presentation-stream),
+static site generators use frontmatter to embed YAML at the top of files.
+Other systems have extended this for non-YAML use, like
+[zola using `+++` for TOML](https://www.getzola.org/documentation/content/page/#front-matter).
+
+## Proposed Python syntax
+
+Currently the draft [PEP 723](https://peps.python.org/pep-0723/) proposes allowing begin/end markers inside of regular comments:
+
+```python
+# /// pyproject
+# [run]
+# requires-python = ">=3.11"
+# dependencies = [
+#   "requests<3",
+#   "rich",
+# ]
+# ///
+
+import requests
+from rich.pretty import pprint
+
+resp = requests.get("https://peps.python.org/api/peps.json")
+data = resp.json()
+pprint([(k, v["title"]) for k, v in data.items()][:10])
+```
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- Support infostring attributes
+  - We need to better understand use cases for how this should be extended, particularly what the syntax should be (see infostring language)
+  - A safe starting point could be to say that a space or comma separates attributes and everything after it is defined as part of the "language"
+- Add support for a `#[frontmatter(info = "", content = "")]` attribute that this syntax maps to.
+  - Since nothing will read this, whether we do it now or in the future will have no affect
+
+## Multiple frontmatters
+
+At least for cargo's use cases, the only other file that we would consider supporting is `Cargo.lock`
+and we have other avenues we want to explore as future possibilities before we
+even consider the idea of multiple frontmatters.
+
+So **if** we decide we need to embed additional metadata, we have a couple of options for extending frontmatter support.
+
+Distinct blocks, maybe with newlines
+
+````md
+```Cargo.toml
+```
+
+```Cargo.lock
+```
+````
+
+Continuous blocks
+````md
+```Cargo.toml
+```Cargo.lock
+```
+````
```

Comment:
`````````suggestion
````md
---Cargo.toml
---

---Cargo.lock
---
````

Continuous blocks
````md
---Cargo.toml
---Cargo.lock
---
````
`````````

Also, the current RFC requires

> Opens with 3+ dashes followed by 0+ whitespace, ***an*** optional identifier, 0+ whitespace, and a newline 

so the infostring can't be `Cargo.toml`/`Cargo.lock` but something like `Cargo_toml`/`Cargo_lock`.

Or just relax the infostring to allow `[^\r\n]+`.

Reply:
I think we should just support infostrings with `.` characters, I'm thinking we support the regex `([^,\r\n],?)*` where each comma-separated item has leading and trailing whitespace trimmed, that way you get a list of items and allows using file names or paths as well as specifying options (e.g. if file permissions (not the main .rs file, but the embedded file) might change something for some tool, there could be a `unix_permissions=0755` option)

Reply:
Thanks!  I updated the code block.  Splitting out the syntax question to a separate thread for easier tracking of distinct items